lib/spec/unit/facter_spec.rb

```diff
@@ -0,0 +1,115 @@
+      version = Facter.value('facterversion')
+      expect(version).to eq(Facter.version)
+      expect(version).to eq(Facter::FACTERVERSION)
+    end
+  end
+
+  describe 'with resetting' do
+    before :each do
+      Facter.reset
+    end
+
+    it 'should load external facts' do
+      # Check for windows vs posix for executable external facts
+      windows = Facter.value('osfamily') == 'windows'
+      Facter.reset
+
+      Facter.search_external([
+        File.expand_path('../../../lib/tests/fixtures/facts/external/yaml', File.dirname(__FILE__)),
+        File.expand_path('../../../lib/tests/fixtures/facts/external/json', File.dirname(__FILE__)),
+        File.expand_path('../../../lib/tests/fixtures/facts/external/text', File.dirname(__FILE__)),
+        File.expand_path("../../../lib/tests/fixtures/facts/external/#{ if windows then 'windows' else 'posix' end }/execution", File.dirname(__FILE__))
+      ])
+
+      facts = Facter.to_hash
+      expect(facts['yaml_fact1']).to be_a(String)
+      expect(facts['yaml_fact2']).to be_a(Integer)
+      expect(facts['yaml_fact3']).to(satisfy { |v| v == true || v == false })
+      expect(facts['yaml_fact4']).to be_a(Float)
+      expect(facts['yaml_fact5']).to be_a(Array)
+      expect(facts['yaml_fact6']).to be_a(Hash)
+      expect(facts['json_fact1']).to be_a(String)
+      expect(facts['json_fact2']).to be_a(Integer)
+      expect(facts['json_fact3']).to(satisfy { |v| v == true || v == false })
+      expect(facts['json_fact4']).to be_a(Float)
+      expect(facts['json_fact5']).to be_a(Array)
+      expect(facts['json_fact6']).to be_a(Hash)
+      expect(facts['exe_fact1']).to be_a(String)
+      expect(facts['exe_fact2']).to be_a(String)
+      expect(facts['exe_fact3']).to be_nil
+      expect(facts['txt_fact1']).to be_a(String)
+      expect(facts['txt_fact2']).to be_a(String)
+      expect(facts['txt_fact3']).to be_nil
+    end
+
+    it 'should set search paths' do
+      Facter.search('foo', 'bar', 'baz')
+      expect(Facter.search_path).to eq(['foo', 'bar', 'baz'])
+      Facter.reset
+      expect(Facter.search_path).to eq([])
+    end
+
+    it 'should set external search paths' do
+      Facter.search_external(['foo', 'bar', 'baz'])
+      expect(Facter.search_external_path).to eq(['foo', 'bar', 'baz'])
+    end
+
+    it 'should support stubbing for confine testing' do
+      Facter.fact(:osfamily).expects(:value).at_least(1).returns 'foo'
+      expect(Facter.fact(:osfamily).value).to eq('foo')
+      Facter.add(:test) do
+        confine osfamily: 'foo'
+        setcode do
+          'bar'
+        end
+      end
+      expect(Facter.value(:test)).to eq('bar')
+    end
+
+    it 'should allow stubbing on which and exec' do
+      Facter::Util::Resolution.expects(:which).with("foo").returns('/usr/bin/foo')
+      Facter::Util::Resolution.expects(:exec).with("foo").returns('bar')
+      expect(Facter::Util::Resolution.which('foo')).to eq('/usr/bin/foo')
```

Comment:
I think the stubbing of `which` was related to the way one fact I looked at was using the facter method... custom facts tend to use `Facter::Util::Resolution` stuff (mostly `exec`, and sometimes `which`) because it work well, which means those are what get stubbed in tests.

`which` and `exec` would both be stubbed if the fact is calling them directly, as @peterhuene pointed out, to "do something correctly."